crates/binstalk-fetchers/src/dist_manifest_fetcher.rs

```diff
@@ -0,0 +1,182 @@
+use std::{
+    collections::BTreeMap,
+    ffi::OsStr,
+    path::{Path, PathBuf},
+};
+
+use binstalk_downloader::remote::{Client, Response};
+use cargo_dist_schema::Format;
+use compact_str::CompactString;
+use normalize_path::NormalizePath;
+
+use crate::FetchError;
+
+/// An tarball/zip artifact.
+#[derive(Clone, Debug)]
+struct BinaryArtifact {
+    /// Filename of artifact on release artifacts,
+    /// need to infer the format.
+    filename: CompactString,
+    /// Path to the executable within the tarbal/zip.
+    path_to_exe: PathBuf,
+
+    /// Filename of the checksum file.
+    checksum_filename: Option<CompactString>,
+}
+
+impl BinaryArtifact {
+    /// Return `Self` and target if the artifact contains binary, or `None`.
+    fn new<'artifacts>(
+        artifact_id: &str,
+        artifacts: &'artifacts BTreeMap<String, cargo_dist_schema::Artifact>,
+        app_name: &str,
+    ) -> Result<Option<(Self, impl Iterator<Item = &'artifacts String>)>, FetchError> {
+        let get_artifact = |artifact_id| {
+            artifacts.get(artifact_id).ok_or_else(|| {
+                FetchError::InvalidDistManifest(format!("Missing artifact_id {artifact_id}").into())
+            })
+        };
+        let binary_artifact = get_artifact(artifact_id)?;
+
+        if !matches!(
+            binary_artifact.kind,
+            cargo_dist_schema::ArtifactKind::ExecutableZip
+        ) {
+            return Ok(None);
+        }
+        let Some(filename) = binary_artifact.name.as_deref() else {
+            return Ok(None);
+        };
```

Comment:
```suggestion
        let (cargo_dist_schema::ArtifactKind::ExecutableZip, Some(filename)) = (binary_artifact.kind, binary_artifact.name.as_deref()) else {
            return Ok(None);
        };
```

Reply:
Thanks, I've applied this advice when rewriting the code.